File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

import java.util.*;

class Solution {
    // DFS
    // public int minDepth(TreeNode root) {
    //     // root 가 null 이면 return 0
    //     if (root == null){
    //         return 0;   
    //     }else if(root.left == null && root.right == null){
    //         return 1;
    //     }else if(root.left == null){
    //         return 1 + minDepth(root.right);
    //     }else if(root.right == null){
    //         return 1 + minDepth(root.left);
    //     }else{
    //         return 1 + Math.min(minDepth(root.left), minDepth(root.right));
    //     }
    // }
    
    // BFS
    public int minDepth(TreeNode root){
        if(root == null){
            return 0;
        }
        
        Queue<TreeNode> q = new LinkedList<>();
        int depth = 0;
        q.offer(root);
        
        while(!q.isEmpty()){
            depth++;
            int size = q.size();
            for(int i = 0; i < size; i++){
                TreeNode node = q.poll();
                
                if(node.left == null && node.right == null){
                    return depth;
                }
                
                if(node.right != null){
                    q.offer(node.right);
                }
                if(node.left != null){
                    q.offer(node.left);
                }
            }
        }
        return depth;
    }
}

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

import queue

class Solution:
    # DFS
#     def minDepth(self, root: Optional[TreeNode]) -> int:
#         if not root:
#             return 0
        
#         left = self.minDepth(root.left)
#         right = self.minDepth(root.right)
        
#         if not left and not right:
#             return 1
        
#         if not left:
#             return 1 + right;
        
#         if not right:
#             return 1 + left;
        
#         return min(left, right) + 1
    
    # BFS
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        
        q = queue.Queue()
        q.put(root)
        depth = 0
        
        while not q.empty():
            size = q.qsize()
            depth += 1
            for num in range(size):
                node = q.get()
                if node.left is None and node.right is None:
                    return depth
                
                if node.right:
                    q.put(node.right)
                    
                if node.left:
                    q.put(node.left)
            
        return depth
        
        